### PR TITLE
XMLDecoder improve attribute name resolution

### PR DESCRIPTION
**Pull request by ghost**
If we aren't able to get the name from the StringPoolChunk we can try falling back to the ResourceMapChunk. This change does add a TODO due to a somewhat naive implementation but still name resolution is improved. ie previously an empty attribute name now shows as attr/name

Ideally we should also resovle **attr/name -> android:name** etc but my knowledge is still limited

Example from one of the Janky tests
[image](https://github.com/Col-E/binary-resources/assets/37674069/2723e2b3-8e4f-4f8a-97da-a8810c9c8c36)

### CODE REVIEW

**Comment by 1fxe** [comment]
Improved the mapping by utilising public.xml from https://android.googlesource.com/platform/frameworks/base/+/57af565fb79c006522261fc211b078d2a5efd331/core/res/res/values/public.xml

image

File: src/main/java/software/coley/android/xml/AndroidResourceProvider.java [hunk 1]
```diff
@@ -61,4 +61,15 @@ public interface AndroidResourceProvider {
 	 */
 	@Nullable
 	String getResEnumName(String resName, long value);
+
+	/**
+	 * @param type
+	 * 		type of resource "attr", "style", "drawable", etc.
+	 * @param resId
+	 * 		Resource ID in the {@code XmlResourceMapChunk}
```
Comment (Col-E) [inline src/main/java/software/coley/android/xml/AndroidResourceProvider.java]: Can use `@link` over `@code`

Comment (ghost) [reply src/main/java/software/coley/android/xml/AndroidResourceProvider.java]: Will do I removed this change as its overkill used a simpler solution

**Comment by ghost** [comment]
I spent an hour coding some public.xml because its 3am simplified the solution after break 🧠 
[image](https://github.com/Col-E/binary-resources/assets/37674069/92d788e9-a751-47a0-a1da-2541b2eabac7)

File: src/main/java/software/coley/android/xml/XmlDecoder.java [hunk 2]
```diff
@@ -182,6 +183,31 @@ public String getReconstructedXml() {
...
+	@Nonnull
+	private String getAttributeName(@Nonnull XmlAttribute attribute) {
+		String name = attribute.name();
+		if (!(name == null || name.isEmpty()))
+			return name;
+
+		if (resourceMap == null)
+			return "";
+
+		BinaryResourceIdentifier resourceId = resourceMap.getResourceId(attribute.nameIndex());
```
Comment (ghost) [inline src/main/java/software/coley/android/xml/XmlDecoder.java]: Might be good to do an index check in getResourceId()

File: src/main/java/software/coley/android/xml/XmlDecoder.java [hunk 3]
```diff
@@ -182,6 +183,31 @@ public String getReconstructedXml() {
...
+		name = resourceProvider.getResName(resourceId.id());
+		// See res-map.txt
```
Comment (Col-E) [inline src/main/java/software/coley/android/xml/XmlDecoder.java]: The `res-map` is an internal implementation provided only by our test module. Recaf will have the same implementation, but a comment here should not refer to it.

Res-map is a flat view of the Android SDK's ARSC bundle's id to name mappings for attributes, created from JADX's tooling.

Comment (ghost) [reply src/main/java/software/coley/android/xml/XmlDecoder.java]: We can use 

https://android.googlesource.com/platform/frameworks/base/+/57af565fb79c006522261fc211b078d2a5efd331/core/res/res/values/public.xml

File: src/main/resources/android/res-map.txt [hunk 4]
```diff
@@ -0,0 +1,12779 @@
+# The base resources can be found at https://android.googlesource.com/platform/frameworks/base/+/refs/heads/main/core/res/res/values/
+# Parsed from android.jar (Android-33) from resources.arsc
```
Comment (ghost) [inline src/main/resources/android/res-map.txt]: android.jar contains all the base resources used by every app, each app can also have its own resources (we handle this already when creating the primary resource)

**Comment by Col-E** [comment]
Why was the `res-map.txt` moved from the test resources to the main resources?

**Comment by ghost** [comment]
> Why was the `res-map.txt` moved from the test resources to the main resources?

I think it should be part of the library resources because every android app has resources from android.jar, the alternative is people would add their own android.jar to get the public resources

**Comment by Col-E** [comment]
That was my intended solution, as then you could update the file independant of this library. For usage in Recaf for instance, I intend to keep our models updated with new API releases. Consumers of the library can reference the implementation in the test module when making their own.

Ideally though, this intent should be explicitly documented on this repo. I can make the documentation for that.

**Comment by ghost** [comment]
> Ideally though, this intent should be explicitly documented on this repo. I can make the documentation for that.

Yeah sounds good ill move it back to test resources and leave the implementation for loading the new res map for another time

**Pull request merged**